### PR TITLE
Thermal optimization clean

### PR DESCRIPTION
**Pull request by Bo-Zhang1995**
Clean the code for optimization, and add two cases for the optimization function.

### CODE REVIEW

**Comment by Xiangyu-Hu** [comment]
@Bo-Zhang1995 when you initialize new action, you can cancel the old no-use actions.

File: src/shared/materials/diffusion_reaction.cpp [hunk 1]
```diff
@@ -4,6 +4,13 @@
 namespace SPH
 {
 //=================================================================================================//
+void LocalIsotropicDiffusion::initializeLocalParameters(BaseParticles* base_particles)
+{
+    base_particles->registerVariable(local_thermal_conductivity_, "ThermalConductivity", [&](size_t i) -> Real {return diff_cf_; });
+    base_particles->addVariableToWrite<Real>("ThermalConductivity");
+    base_particles->addVariableToRestart<Real>("ThermalConductivity");
```
Comment (Xiangyu-Hu) [inline src/shared/materials/diffusion_reaction.cpp:11]: material property do not need for restart?

File: src/shared/materials/diffusion_reaction.h [hunk 2]
```diff
@@ -81,12 +82,41 @@ class IsotropicDiffusion : public BaseDiffusion
...
+    virtual ~LocalIsotropicDiffusion() {};
+
+    virtual void initializeLocalParameters(BaseParticles* base_particles) override;
+
+    virtual Real getReferenceDiffusivity() override { return diff_cf_; };
+    virtual Real getDiffusionCoeffWithBoundary(size_t particle_i) override { return local_thermal_conductivity_[particle_i]; };
+    virtual Real getInterParticleDiffusionCoeff(size_t particle_i, size_t particle_j, const Vecd& direction_from_j_to_i) override
+    {
+        return 0.5 * (local_thermal_conductivity_[particle_i] + local_thermal_conductivity_[particle_j]);
+    };
```
Comment (Xiangyu-Hu) [inline src/shared/materials/diffusion_reaction.h:117]: consider to use template for local diffusion property?

File: src/shared/particle_dynamics/diffusion_optimization_dynamics/diffusion_optimization.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+* and HU1527/12-1.															*
+*                                                                           *
+* Portions copyright (c) 2017-2020 Technical University of Munich and		*
+* the authors' affiliations.												*
+*                                                                           *
+* Licensed under the Apache License, Version 2.0 (the "License"); you may   *
+* not use this file except in compliance with the License. You may obtain a *
+* copy of the License at http://www.apache.org/licenses/LICENSE-2.0.        *
+*                                                                           *
+* --------------------------------------------------------------------------*/
```
Comment (Xiangyu-Hu) [inline src/shared/particle_dynamics/diffusion_optimization_dynamics/diffusion_optimization.h:22]: use new credit format without using tabs.

**Comment by Xiangyu-Hu** [comment]
@Bo-Zhang1995 better put together with some kind of test.

**Comment by Xiangyu-Hu** [comment]
> @Bo-Zhang1995 better put together with some kind of test.

Sorry. I have found them.

**Review by Xiangyu-Hu — APPROVED** [review]

**Comment by ChiZhangatTUM** [comment]
@Bo-Zhang1995  0.5(k1 + k2) is not an accurate approximation for material with different conductivity, you may try 2.0 * k1 * k2 / (k1 + k2).  I am trying to rewrite the diffusion part, you can keep this in mind.

**Comment by Bo-Zhang1995** [comment]
> @Bo-Zhang1995 0.5(k1 + k2) is not an accurate approximation for material with different conductivity, you may try 2.0 * k1 * k2 / (k1 + k2). I am trying to rewrite the diffusion part, you can keep this in mind.

Thanks for your information. 0.5(k1 + k2) approximation is easier for the evolution of k during the current optimization because it can be written as 0.5(k1+dk1+k2+dk2). I will try to explore the evolution scheme with 2.0 * k1 * k2 / (k1 + k2) for further optimization study. Thanks.

**Comment by ChiZhangatTUM** [comment]
> > @Bo-Zhang1995 0.5(k1 + k2) is not an accurate approximation for material with different conductivity, you may try 2.0 * k1 * k2 / (k1 + k2). I am trying to rewrite the diffusion part, you can keep this in mind.
> 
> Thanks for your information. 0.5(k1 + k2) approximation is easier for the evolution of k during the current optimization because it can be written as 0.5(k1+dk1+k2+dk2). I will try to explore the evolution scheme with 2.0 * k1 * k2 / (k1 + k2) for further optimization study. Thanks.

Understood.

**Pull request merged**